Py2to3

madhuBayy opened the pull request:
@ldko  @somexpert 
This MR  upgrades python2 code to python3. Please review the changes.

madhuBayy committed:
Port the code to python3

madhuBayy committed:
Update python version in travis.yml

madhuBayy committed:
Update pyuntl dependency to the latest version.

madhuBayy committed:
Rename map function.

madhuBayy committed:
Update python versions in setup.py, travis.yml

ldko commented on m2m/m2m.py:
```diff
@@ -39,11 +39,6 @@ def CSVToDict(csvFileName):
-            try:
-                d[k] = unicode(v, 'utf-8', 'replace')  # probably not smart
-            except:
-                pass  # not sure how smart /this/ is either...
```
If we no longer need a unicode conversion on the values, would it be the same to replace lines 40 onward with:
`return list(readerDict)`?

madhuBayy replied:
Oops..I misread your comment. I can convert it to a list directly as you suggested.

ldko commented on m2m/m2m.py:
```diff
@@ -160,8 +155,8 @@ def writeTemplateFiles(self, baseDirectory, foldername):
+        templateFile = open(os.path.join(writeDirectory, "metadata.xml"), "wb")
+        xmlFile = self.__bytes__()
         templateFile.write(xmlFile)
         templateFile.close()
```
Perhaps we should change this to:
```
        with open(os.path.join(writeDirectory, "metadata.xml"), "wb") as templateFile:
            templateFile.write(self.__bytes__())
```

ldko commented on m2m/m2m.py:
```diff
@@ -71,7 +66,7 @@ def setBaseDirectory(self, BaseDirectory):
 
-    def map(self, elementType, elementName, elementValue, qualifier=None,
+    def mapping(self, elementType, elementName, elementValue, qualifier=None,
             required=True, info="", location="", agent_type="", split="",
```
Not something you introduced, but down below, ` if elementType is not "basic" and elementType is not "agent":` should be changed to use `!=` or be changed to `if elementType not in {"basic", "agent")`

ldko commented on setup.py:
```diff
@@ -4,7 +4,7 @@
     name='m2m',
-    version='1.0.0',
+    version='2.0.0',
     url='https://github.com/unt-libraries/m2m',
```
perhaps lower down we should also add:
```
    dependency_links=[
        'git+https://github.com/unt-libraries/pyuntl.git/@master#egg=pyuntl-2.0.0'
    ],
```

ldko reviewed (commented):
In addition to the comments I have inline, the example code in the README should be updated for Python 3 and the `mapping` change.

madhuBayy committed:
Update Readme code, and make some minor code changes.

ldko commented on requirements.txt:
```diff
@@ -1 +0,0 @@
-pyuntl
```
I think we could have kept this file with the change you had introduced earlier to get the correct version. I think it is ok to have both this and the dependency in setup.py

ldko replied:
Oh also, .travis.yml expects this file, so the build is failing.

madhuBayy replied:
yea..that's why I was wondering should I bring this file back or if I can use git dependency link in .travis.yml

ldko commented on README.md:
```diff
@@ -4,11 +4,11 @@ m2m is a python package and command line tool to assist in converting csv docume
+>>> record.mapping("basic", "title", "Pawn of Prophecy", qualifier="officialtitle")
+>>> record.mapping("agent", "creator", "Eddings, David", qualifier="aut", agent_type="per")
+>>> record.mapping("basic", "date", "1982", qualifier="creation")
+>>> print(bytes(record))
```
I wonder if we should bring back a `__str__` method, so we don't need to do the `bytes` call for `print`. Something like:
```
    def __str__(self):
        return self.__bytes__().decode()
```
What do you all think @madhulika95b @vphill @somexpert ?

madhuBayy replied:
I think it'd be nice to have `__str__`, but I think we still have to call `str(record)` to print.
Edit: Ah, I see, `print()` automatically calls `__str__()`.

madhuBayy committed:
Add requirements file back.

vphill committed:
Adding test mapping file.

This can be run with the `m2m.py` command. 

Here is what I would expect to get as output from the command.

```bash
mphillips@LSLDW370134434 test % python3 m2m/m2m/m2m.py -m test_2_untl.py test.csv
Processing CSV file test.csv with mapping test_2_untl.py
Processing row 0
<?xml version="1.0" encoding="UTF-8"?>
<metadata>
  <title qualifier="officialtitle">Graceling</title>
  <creator qualifier="aut">
    <name>Kristin Cashore</name>
    <info>born somewhere</info>
    <type>per</type>
  </creator>
  <date qualifier="creation">2008</date>
  <meta qualifier="metadataCreator">mphillips</meta>
</metadata>
```

madhuBayy committed:
Add setup instructions in README.

madhuBayy committed:
Add tox.ini and ignore build, dist, .tox, .egg dirs.

madhuBayy committed:
Add __str__().

ldko commented on tox.ini:
```diff
@@ -0,0 +1,15 @@
+[flake8]
+max-line-length = 99
+
+
```
Seems like we could remove one of these extra blank lines.

ldko commented on README.md:
````diff
@@ -20,9 +20,39 @@ m2m is a python package and command line tool to assist in converting csv docume
 ```
 
-# License
+Requiremnets
````
Should be `Requirements`.

ldko commented on README.md:
```diff
@@ -20,9 +20,39 @@ m2m is a python package and command line tool to assist in converting csv docume
+Installation
+--------------
+
+This application can be installed by following the steps below:
```
Maybe "can be installed and run" since you have a command to run it as well as install.

ldko commented on tox.ini:
```diff
@@ -0,0 +1,15 @@
+
+[testenv:py37-flake8]
+deps=flake8 
+commands=flake8 m2m tests setup.py
```
Could add spaces around `=` for consistency.

ldko reviewed (commented):
flake8 checking was added in .tox, but, as that is not used in .travis.yml,, we aren't checking flake8 in CI. Also flake8 checking fails with running tox locally, so I am wondering if that should be added later when the issue about pep8 is addressed? Though I don't feel strongly about that...What do you think?

madhuBayy committed:
Fix typos and remove flake8 check for now.

ldko commented on tox.ini:
```diff
@@ -0,0 +1,10 @@
+
+[flake8]
+max-line-length = 99
+
```
We don't really need this flake8 section here after having removed the other flake8 stuff, though it probably doesn't hurt anything to stay if you are going to add flake8 check back soon in another PR.

vphill committed:
Update test.csv

adding isbn column

vphill committed:
Update test_2_untl.py

adding example usage of `setBaseDirectory` and `setFolderName`

madhuBayy committed:
Remove flake8 check from tox.

somexpert commented on README.md:
```diff
@@ -20,9 +20,39 @@ m2m is a python package and command line tool to assist in converting csv docume
+
+    $ python setup.py install
+
+    $ python m2m/m2m.py tests/data/test.csv -m tests/data/test_2_untl.py
```
I feel like this should be in an "example" or have a comment showing this is how it can be used, rather than just listed as another "step" in the installation process.

madhuBayy replied:
sure. I'll add a comment.

somexpert commented on README.md:
```diff
@@ -20,9 +20,39 @@ m2m is a python package and command line tool to assist in converting csv docume
+
+To run the development tests, use the following command:
+
+    $ tox
```
Now that we're using tox, can we get rid of runtests.py?

madhuBayy replied:
yes, I think so. Will remove it.

somexpert commented on m2m/m2m.py:
```diff
@@ -57,28 +49,31 @@ class MetadataRecord(object):
         """
 
-        if elementType is not "basic" and elementType is not "agent":
+        if elementType not in {"basic", "agent"}:
```
Is there a reason to use a set here instead of a list or tuple?

madhuBayy replied:
nothing particular honestly.

somexpert replied:
I would personally change it to a list or tuple then, though the only real reason I have for it is for consistency with how we typically do this in other repos.

somexpert reviewed (commented):
Just a few comments.

somexpert commented:
~~I'm also beginning to question whether requirements.txt really serves a purpose still or not.~~ nvm, forgot about tox

madhuBayy commented:
well, I think we can get rid of requirements.txt file and add code coverage to tox and run tests in .travis.yml using tox. what do you say @ldko ?

somexpert commented:
well the problem is that tox uses pip for installation, and pip doesn't use the dependency_links, which means without installing from requirements.txt it grabs the wrong version of pyuntl. I have a solution that I'd like to propose, that will allow us to eliminate that requirements file and still be able to use tox and install with both pip and setup.py (currently pip won't work), but give me a bit to formulate it properly

madhuBayy commented:
We can use the pyuntl master branch, the way we are using in requirements.txt file and it gets the correct version in tox.

somexpert commented on setup.py:
```diff
@@ -4,22 +4,28 @@
     install_requires=['pyuntl'],
+    dependency_links=[
+        'git+https://github.com/unt-libraries/pyuntl.git/@master#egg=pyuntl-2.0.0'
+    ],
```
So currently, this can be successfully installed with `python setup.py install`, which reads the `dependency_links`, but not with `pip install .`, which does not. Since we use tox, which installs the package with `pip`, we currently use `requirements.txt` to ensure we grab the correct version of `pyuntl`. That still doesn't allow us to install with `pip install .`, and while that isn't necessary, I propose the following changes, which will allow for both installation methods _and_ allow us to get rid of `requirements.txt` and the related installation in `tox.ini`:

define the URL as a variable above `setup()`:
`pyuntl_url = 'git+https://github.com/unt-libraries/pyuntl.git@master#egg=pyuntl`

then change `install_requires` like so (pip, and by extension, tox, but _not_ setup.py, will see the link here and use it):
`install_requires=['pyuntl @' + pyuntl_url]`

and finally, just to remove some duplication:
`dependency_links=[pyuntl_url]`

Now, is this the _best_ method? I kinda doubt it. But I have yet to find something definitive on how best to solve the issue of dependencies outside of PyPI that works with both setup.py and pip. Just for some context, here's a couple SO posts that I keep seeing whenever I try looking into this further:
https://stackoverflow.com/questions/26061610/pip-install-dependency-links
https://stackoverflow.com/questions/32688688/how-to-write-setup-py-to-include-a-git-repo-as-a-dependency

ldko replied:
Sound like an ok thing to try to me.

madhuBayy replied:
Works fine. And.. doesn't need pyuntl installation separately in tox either.

madhuBayy committed:
Remove requirements.txt, runtests.py files.

madhuBayy committed:
Use a url for pyuntl dependency link and update readme instructions.

madhuBayy committed:
Ignore .coverage files.

madhuBayy committed:
Use tuple to check elementType membership.

madhuBayy committed:
Add command to install dependencies in travis build.

madhuBayy committed:
Fix travis build by adding python envs.

somexpert commented on README.md:
```diff
@@ -28,13 +28,15 @@ Requirements
     $ cd m2m
 
-    $ python setup.py install
+    $ pip install .
```
You didn't have to change this, btw. It should now work fine with _either_ installation method. Use whichever you want

somexpert commented on .travis.yml:
```diff
@@ -5,6 +5,7 @@ python:
   - '3.8'
 # command to install dependencies
 install:
+  - pip install .
```
I don't think this is needed. Tox will install the package itself.

madhuBayy replied:
sure! Got rid of it.

somexpert commented:
I _think_ the remaining Travis issue might be related to pip version. Maybe try
`pip install --upgrade pip`
in the Travis setup and see if that resolves the issue?

madhuBayy committed:
Upgrade pip version in travis build.

madhuBayy commented:
Hmm...upgrading pip which was previously `19.0.3` and now after upgrading `20.1.1`, didn't seem to solve the issue. I'm wondering if this is something with python 3.6?

madhuBayy committed:
set tox config to upgrade pip/setuptools.

somexpert commented on .travis.yml:
```diff
@@ -9,7 +9,7 @@ matrix:
 # command to install dependencies
 install:
-  - pip install .
+  - pip install --upgrade pip
```
we can get rid of this now

madhuBayy committed:
Remove pip upgrade from travis.

somexpert reviewed (approved):
LGTM :thumbsup:
